DOC: signal.detrend: clarify `overwrite_data` behaviour

OleBialas opened the pull request:
The argument `overwrite_data=True` made it appear as if overwriting was guaranteed. However, overwriting only happens when ``type == linear`` and ``dtype is in "dtDF"`` and array reshaping returns a view. I changed the argument name to `allow_overwrite` and updated the docstring to indicate overwriting only happens under specific conditions.

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-14498

#### What does this implement/fix?
<!--Please explain your changes.-->
This makes the behavior for in place modifications more explicit

#### Additional information
<!--Any additional information you think is important.-->

OleBialas commented:
@Kai-Striega

lucascolley commented:
Thanks @OleBialas !

Would you mind splitting the formatting changes into a separate commit? Then we can add it to https://github.com/scipy/scipy/blob/main/.git-blame-ignore-revs to hide them from git blame. It also makes it easier to review the real changes here.

OleBialas committed:
DOC:signal._signaltools.detrend: Rename overwrite_data.

The argument `overwrite_data=True` made it appear as if overwriting was
guaranteed. However, overwriting only happens when ``type == linear``
and ``dtype is in "dtDF"`` and array reshaping returns a view.
I changed the argument name to `allow_overwrite` and updated the
docstring to indicate overwriting only happens under specific
conditions.

lucascolley reviewed (commented):
I think we will need to introduce a deprecation cycle, where the old parameter name is kept around as an alias for the new one, but raises a deprecation warning if used. There is an example of a similar change at https://github.com/scipy/scipy/pull/16004/files.

I think we should raise an exception if both `overwrite_data` and `allow_overwrite` are passed with different values.

We probably also want to keep the `overwrite_data` parameter in the docstring, but add a `.. deprecated:: 1.17.0` admonition, like those seen at https://github.com/scipy/scipy/commit/ca34a7735cae547568ef88729432ebaed7403275.

j-bowhay reviewed (dismissed):
Thanks for your contribution @OleBialas!

I agree if starting from scratch that `allow_overwrite` would be a better name. Unfortunately, we cannot proceed with your proposed renaming as it would break backwards compatibility. I would propose keeping the old name and improving the documentation (as you have done)

lucascolley commented:
I'm also open to Jake's suggestion of just improving the documentation.

OleBialas commented:
I'm fine with either introducing the deprecation warning or updating only the documentation

j-bowhay commented:
I would prefer just to update the documentation as this stage as a deprecation is a lot more effort. Then if someone still finds it confusing in the future we can go through that whole process

OleBialas committed:
only update docstring, not variable name

OleBialas committed:
remove _NoValue

OleBialas committed:
add back typehint

OleBialas committed:
update type hint

OleBialas commented:
I was actually wondering why overwriting is completely ignored for constant detrending. For now, I updated my PR to only change the documentation, but I would be open to work on the suggestion by @DietBru in a separate issue

lucascolley committed:
typo

[skip ci]

lucascolley reviewed (commented):
<img width="648" height="116" alt="image" src="https://github.com/user-attachments/assets/041bd585-5e82-4bd2-82c8-7e2f45c3fd85" />

preview looks good!

> overwriting only happens when ``type == linear`` and ``dtype is in "dtDF"`` and array reshaping returns a view.

maybe we should spell out these last two points in the Notes section? What do you think @DietBru @j-bowhay ?

DietBru commented:
Yup. Might be useful to note that overwriting only works for floats. In a follow-up PR overwriting could be also made workable for complex-valued inputs.

OleBialas committed:
Update scipy/signal/_signaltools.py

Co-authored-by: Dietrich Brunn <12721170+DietBru@users.noreply.github.com>

OleBialas commented on scipy/signal/_signaltools.py:
```diff
@@ -4072,8 +4072,10 @@ def detrend(data: np.ndarray, axis: int = -1,
+    overwrite_data: bool, optional
+        If True, allow in place detrending and avoid a copy. Default is
+        False. In place modification applies only if ``type == 'linear'``
+        and also depends on the data type and memory layout of `data`.
```
Looks good - the more specific the documentation, the better!

DietBru reviewed (approved):
The [rendered docstr] looks good to me.

[rendered docstr]: https://output.circle-artifacts.com/output/job/e0037579-9457-4e1e-a505-c8fc98ab5edd/artifacts/0/html/reference/generated/scipy.signal.detrend.html

lucascolley reviewed (approved):
Also LGTM, thanks all!
